src/test/Spec.hs

```diff
@@ -0,0 +1,120 @@
+        --   case call1 of
+        --        Left ("set0", [AbiInt _ 0]) -> True
+        --        _                           -> False
+        --   &&
+        --   case call2 of
+        --        Left ("set1", [AbiInt _ 0]) -> True
+        --        _                           -> False
+  , testCase "Optimal Solve" $ testContract' c3 $
+      \c -> do
+        let tr = findtest' c "echidna_revert"
+        assertBool "echidna_revert unsolved" $ solved tr
+        let sol = solve tr
+        assertBool "solution has length /= 1" $ length sol == 1
+        let sol' = head sol
+        assertBool "solution is not f(-1)" $
+          case sol' ^. call of
+               Left ("f", [AbiInt _ (-1)]) -> True
+               _                           -> False
+  {-, testCase "Payment amounts" $ testContract c4 $
+      \c -> do
+        let tr = findtest' c "echidna_test"
+        assertBool "echidna_test unsolved" $ solved tr -}
+  , testCase "Multisender" $ testContract c6 (Right cfg6) $
+      \c -> do
+        let tr = findtest' c "echidna_all_sender"
+        assertBool "echidna_all_sender unsolved" $ solved tr
+        let sol = solve tr
+        assertBool "solution has length /= 3" $ length sol == 3
+        let calls = view call <$> sol
+        assertBool "s1 not in solution" $ any (\case {Left ("s1", _) -> True; _ -> False;}) calls
+        assertBool "s2 not in solution" $ any (\case {Left ("s2", _) -> True; _ -> False;}) calls
+        assertBool "s3 not in solution" $ any (\case {Left ("s3", _) -> True; _ -> False;}) calls
+  ]
+  where c2   = "./examples/solidity/basic/flags.sol"
+        c3   = "./examples/solidity/basic/revert.sol"
+        --c4   = "./examples/solidity/basic/payable.sol"
+        c5   = "./examples/solidity/basic/true.sol"
+        c6   = "./examples/solidity/basic/multisender.sol"
+        cfg6 = "./examples/solidity/basic/multisender.yaml"
+
+testContract :: FilePath -> Either EConfig FilePath -> (Campaign -> HU.Assertion) -> HU.Assertion
+testContract file cfg f = do
+  cfg' <- case cfg of
+               Left c   -> pure c
+               Right fp -> maybe (pure defaultConfig) parseConfig (Just fp)
+  let c = cfg' & sConf . quiet .~ True
+  results <- flip runReaderT c $ do
+    (v, w, ts) <- loadTesting file Nothing
+    campaign (pure ()) v w ts
+  f results
+
+testContract' :: FilePath -> (Campaign -> HU.Assertion) -> HU.Assertion
+testContract' = flip testContract (Left defaultConfig)
+
+solved :: TestState -> Bool
+solved (Large _ _) = True
+solved (Solved _)  = True
+solved _           = False
+
+passed :: TestState -> Bool
+passed Passed = True
+passed _      = False
+
+solve :: TestState -> [Tx]
+solve (Large _ s) = s
+solve (Solved s) = s
+solve _ = undefined
+
+findtest :: Campaign -> Text -> Maybe TestState
+findtest = findtest'' . (^. tests)
+
+findtest' :: Campaign -> Text -> TestState
+findtest' = (fromJust .) . findtest
```

Comment:
I'm not sure I like these names/this organization of logic. Let's consolidate this into one function, and avoid partial stuff (`fromJust`) when we can. IMO, it's even better to throw our own `error` b/c that's much more transparent to non-Haskell-familiar users

Reply:
Done (@japesinator )